docs/blueprints/storage-node-downtime-tracking-with-audits.md

```diff
@@ -0,0 +1,153 @@
+# Storage Node Downtime Tracking With Audits
+
+## Abstract
+
+This document describes a means of tracking storage node downtime with audits and using this information to suspend and disqualify.
+
+## Background
+
+The previous implementation of uptime reputation consisted of a ratio of online audits to offline audits. We encountered a problem where some nodes' reputations would quickly become destroyed over a relatively short period of downtime due to the frequency of auditing any particular node being directly correlated with the number of pieces it holds. To solve this problem we need a system that takes into account not only how many offline audits occur, but _when_ they occur as well.
+
+## Design
+
+The solution proposed here is to use a series of sliding windows to indicate a general timeframe in which offline audits occur. Each window keeps two separate tallies indicating how many offline audits and total audits a particular node received within its timeframe. Once a window is complete, it is scored by calculating the percentage of total audits for which it was offline. We can average these scores over a trailing period of time, called the _tracking period_, to determine an overall "offline score" to be used for suspension and disqualification. By granting each individual window the same weight in the calculation of the overall average, the effect of any particularly unlucky period can be minimized while still allowing us to take the failures into account over a longer period.
+
+Storage node downtime can have a range of causes. For those storage node operators who may have fallen victim to a temporary issue, we want to give them a chance to diagnose and fix it before disqualifying them for good. For this reason, we are introducing suspension as a component of disqualification.
+
+Once a node's offline score has risen above an _offline threshold_, it is _suspended_ and enters a period of review. A suspended node will not receive any new pieces, but can continue to receive download and audit requests for the pieces it currently holds. However, its pieces are considered to be unhealthy. We repair a segment if it contains too many unhealthy pieces, at which point we may transfer the repaired pieces from a suspended node to a more reliable node. If at any point during the review period we find that a node's score has fallen below the offline threshold, it is unsuspended, or _reinstated_, but it remains _under review_. This prevents nodes from alternating between suspension and reinstatement without consequence.
```

Comment:
So "reinstated" means that a node's offline score has returned to an acceptable value. When this happens, it is no longer suspended and is considered to be healthy. However, it still remains under review. The reason for this is that, we want to allow nodes to come back ASAP if we see that their online score is okay, but we still want to keep an eye on them. The under review period just means that we're going to check up on them in the future to make sure they're still reliable, otherwise they get DQd

Reply:
Ok, that sounds good - but eventually they will exit under_review, correct? And when they do, they are no longer considered "reinstated", but "healthy"? (assuming they are online)

Reply:
Correct! Once the under_review period is over, the node is either just back to being a happy normal node or it's disqualified

Reply:
Ok, that makes sense to me. Resolving.